mini-projects/01-device-driver/laptop-interface/src/libserial/list_ports.rs

```diff
@@ -0,0 +1,75 @@
+use super::{
+    ffi::{sp_get_port_name, sp_list_ports, SpPort, SpReturn},
+    types::Port,
+};
+use anyhow::Result;
+use std::ffi::CStr;
+use std::ptr;
+
+// Example output:
+// ---- list_ports::tests::test_list_ports stdout ----
+// Port name: /dev/cu.Bluetooth-Incoming-Port
+// Port name: /dev/cu.usbmodem2101
+pub fn list_ports() -> Vec<Port> {
+    unsafe {
+        let mut port_list: *mut *mut SpPort = ptr::null_mut();
+        let result: SpReturn = sp_list_ports(&mut port_list);
+
+        let mut ports: Vec<Port> = Vec::new();
+
+        match result {
+            SpReturn::SpOK => {
+                let mut i = 0;
+                while !(*port_list.add(i)).is_null() {
+                    let port = *port_list.add(i);
+                    let port_name = sp_get_port_name(port);
+                    if !port_name.is_null() {
+                        let port_name_str = CStr::from_ptr(port_name).to_string_lossy();
+                        ports.push(Port {
+                            name: port_name_str.to_string(),
+                            handle: port,
+                        });
+                        println!("Port name: {}", port_name_str);
+                    }
+                    i += 1;
+                }
+            }
+            _ => {
+                println!("Error: {:?}", result);
+            }
+        }
+        ports
+    }
+}
+
+/// Finds the RPi port by name matching the first device containing "cu.usbmodem"
+pub fn get_rpi_port() -> Result<Port> {
```

Comment:
very out of scope for what the mp was expecting, but some food for thought: it might be nice to make the port scanning functionality more robust, and also the init handshake - I can imagine that there would be problems if I had another rpi connected to my laptop for whatever reason

the other main thing I see is when I had initially gone to run your code, I had forgotten to flash the rpi, and so the initialization sequence completed as expected with my default firmware running, but the buttons weren't being processed since you had changed that interface, but no errors were being propagated to the user